src/stages/stage5/timeline.c

```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
 
-	FROM_TO(400, 600, 10) {
-		tsrand_fill(2);
-		create_enemy3c(200.0*I*afrand(0), 500, Swirl, stage5_swirl, 4+I, 70+20*afrand(1)+200.0*I, cexp(-0.05*I));
-	}
+	e->move = move_towards(e->pos + ARGS.acceleration * 100, 0.05);
+	WAIT(100);
 
-	FROM_TO(700, 800, 10) {
-		tsrand_fill(3);
-		create_enemy3c(VIEWPORT_W+200.0*I*afrand(0), 500, Swirl, stage5_swirl, -4+afrand(1)*I, 70+20*afrand(2)+200.0*I, cexp(0.05*I));
-	}
+	int delay = difficulty_value(6, 5, 4, 3) + ARGS.reduction;
+	int amount = 700 / delay;
 
-	FROM_TO(870+50*(global.diff==D_Easy), 1000, 50)
-		create_enemy1c(VIEWPORT_W/4+VIEWPORT_W/2*(_i&1), 2000, BigFairy, stage5_limiter, I);
+	for (int x = 0; x < amount; x++) {
+		cmplx n = cdir(carg(global.plr.pos - e->pos) + (0.2 - 0.02 * global.diff) * x);
+		float fac = (0.5 + 0.2 * global.diff);
 
-	AT(1000)
-		create_enemy1c(VIEWPORT_W/2, 9000, BigFairy, stage5_laserfairy, 2.0*I);
+		// TODO: is this the correct "modern" way of invoking lasers?
+		create_lasercurve2c(e->pos, 100, 300, RGBA(0.7, 0.3, 1, 0), las_accel, fac * 4 * n, fac * 0.05 * n);
 
-	FROM_TO(1400+100*(D_Lunatic - global.diff), 2560, 60-5*global.diff) {
-		tsrand_fill(2);
-		create_enemy1c(VIEWPORT_W+200.0*I*afrand(0), 500, Swirl, stage5_miner, -3+2.0*I*afrand(1));
+		PROJECTILE(
+			.proto = pp_plainball,
+			.pos = e->pos,
+			.color = RGBA(0.7, 0.3, 1, 0),
+			.move = move_accelerated(fac * 4 * n, fac * 0.05 * n),
+		);
+		play_sfx_ex("shot_special1", 0, true);
+		WAIT(delay);
 	}
 
-	FROM_TO(1500, 2400, 80)
-		create_enemy1c(VIEWPORT_W*(_i&1)+100.0*I, 300, Fairy, stage5_greeter, 3-6*(_i&1));
+	WAIT(30);
+	e->move = move_linear(-ARGS.acceleration);
+}
 
-	AT(2500) {
-		create_enemy1c(VIEWPORT_W/2, 2000, BigFairy, stage5_lightburst, 2.0*I);
+TASK(lightburst_fairies_1, {
+	int num;
+	cmplx pos1;
+	cmplx pos2;
+	cmplx exit;
+}) {
+	for (int i = 0; i < ARGS.num; i++) {
+		cmplx pos = ARGS.pos1 + ARGS.pos2 * i;
+		INVOKE_TASK(lightburst_fairy_1,
+			.pos = pos,
+			.move_enter = move_towards(pos + ARGS.exit * 70 , 0.05),
+			.move_exit = move_linear(ARGS.exit)
+		);
+		WAIT(40);
 	}
+}
 
-	FROM_TO(2200, 2600, 60-8*global.diff)
-		create_enemy1c(VIEWPORT_W/(6+global.diff)*_i, 200, Swirl, stage5_miner, 3.0*I);
-
-	AT(2700) {
-		if(global.diff > D_Easy) {
-			create_enemy1c(VIEWPORT_W-20+120*I, 2000, BigFairy, stage5_lightburst, -2.0);
-		}
+TASK(lightburst_fairies_2, {
+	int num;
+	cmplx pos1;
+	cmplx pos2;
+	cmplx exit;
+}) {
+	for (int i = 0; i < ARGS.num; i++) {
+		cmplx pos = ARGS.pos1 + ARGS.pos2 * i;
+		INVOKE_TASK(lightburst_fairy_2,
+			.pos = pos,
+			.move_enter = move_towards(pos + ARGS.exit * 70 , 0.05),
+			.move_exit = move_linear(ARGS.exit)
+		);
+		WAIT(40);
 	}
+}
 
-	AT(2900)
-		global.boss = stage5_spawn_midboss();
+TASK(loop_swirl_move, {
+    MoveParams *move;
+	cmplx turn;
+	cmplx retention;
+}) {
+	int begin = creal(ARGS.turn) + 1;
+	int end = cimag(ARGS.turn) - 1;
+	int time = end - begin;
+
+	ARGS.move->retention = ARGS.retention;
+	WAIT(time / 1.5);
```

Comment:
Why `/ 1.5`?

Reply:
It's the only way to get the swirls to end their "looping" motion at the correct angle. I'm not too sure why, as I was mostly roughing it out at the time.

Reply:
That's not the right way to go about fixing it. You should either change the rotation angle-per-frame (the argument to `cdir()` used for `retention`), or adjust the begin/end timing at call site. There are also a couple more problems with this code that I haven't noticed at first:

* First of all there's no reason to express the timings as a single `cmplx` parameter. The only reason the old code does it is because in the crappy old system you could only have `cmplx` args (and only 4 of them at that). You should just make `int end, begin` task parameters; I'd name them `turn_begin/turn_end` for clarity.
* The +/-1 offsets should be just moved to the call site.
* The `begin` timing isn't used properly. You should `WAIT(begin)` before changing retention to a rotating one. And then you shouldn't use `INVOKE_SUBTASK_DELAYED` - set the `begin` timing properly instead.
* Does the `WAIT()` even work? It seems like you're passing a negative number into it:
    * You are calling the `loop_swirls` task with `turn = 9`. This is looks very wrong...
    * `begin == creal(9) + 1 == 9 + 1 == 10`
    * `end == cimag(9) - 1 == 0 - 1 == -1`
    * `time == end - begin == -1 - 10 == -11`
    * `WAIT(time / 1.5) = WAIT(-7)` - honestly not even sure what happens in this case. We probably should have an assertion for wait time >= 0 inside that macro.


Once you get this all figured out (and preferably *not* before you do), I'd recommend switching to a more intuitive set of outermost parameters for turning:
* `int turn_delay` - time to wait before initiating the turn, in frames.
* `int turn_duration` - duration of the turn, in frames.
* `real turn_angle` - total angle of the turn, in radians.

This is exactly what the stage 2 fairy turning code does, and perhaps it makes sense to generalize and move it to `common_tasks.c` eventually.

Reply:
Done and done. I was worried the movement was going to be to different but wiping out the old format and using what was in Stage 2 ended up being much better. I also moved it to `common_tasks`. I did a before/after comparison and the movement in Stage 2 seemed identical, so there you go.